staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go

```diff
@@ -0,0 +1,272 @@
+
+// isEmptyUnstructuredObject returns true of in is an empty unstructured object, i.e. an unstructured object that does
+// not have any field except apiVersion and kind.
+func isEmptyUnstructuredObject(in runtime.Object) bool {
+	Object, ok := in.(*unstructured.Unstructured)
+	if !ok {
+		return false
+	}
+	if len(Object.Object) != 2 {
+		return false
+	}
+	if _, ok := Object.Object["kind"]; !ok {
+		return false
+	}
+	if _, ok := Object.Object["apiVersion"]; !ok {
+		return false
+	}
+	return true
+}
+
+func (webhookConverter) ConvertFieldLabel(gvk schema.GroupVersionKind, label, value string) (string, string, error) {
+	return "", "", errors.New("unstructured cannot convert field labels")
+}
+
+func (c *webhookConverter) Convert(in, out, context interface{}) error {
+	unstructIn, ok := in.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("input type %T in not valid for unstructured conversion", in)
+	}
+
+	unstructOut, ok := out.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("output type %T in not valid for unstructured conversion", out)
+	}
+
+	outGVK := unstructOut.GroupVersionKind()
+	if !c.validVersions[outGVK.GroupVersion()] {
+		return fmt.Errorf("request to convert CR from an invalid group/version: %s", outGVK.String())
+	}
+	inGVK := unstructIn.GroupVersionKind()
+	if !c.validVersions[inGVK.GroupVersion()] {
+		return fmt.Errorf("request to convert CR to an invalid group/version: %s", inGVK.String())
+	}
+
+	unstructOut.SetUnstructuredContent(unstructIn.UnstructuredContent())
+	_, err := c.ConvertToVersion(unstructOut, outGVK.GroupVersion())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func createConversionReview(obj runtime.Object, apiVersion string) *v1beta1.ConversionReview {
+	listObj, isList := obj.(*unstructured.UnstructuredList)
+	var objects []runtime.RawExtension
+	if isList {
+		for _, item := range listObj.Items {
+			objects = append(objects, runtime.RawExtension{Object: &item})
+		}
+	} else {
+		objects = []runtime.RawExtension{{Object: obj}}
+	}
+	return &v1beta1.ConversionReview{
+		Request: &v1beta1.ConversionRequest{
+			Objects:           objects,
+			DesiredAPIVersion: apiVersion,
+			UID:               uuid.NewUUID(),
+		},
+		Response: &v1beta1.ConversionResponse{},
+	}
+}
+
+func getRawExtensionObject(rx runtime.RawExtension) (runtime.Object, error) {
+	if rx.Object != nil {
+		return rx.Object, nil
+	}
+	u := unstructured.Unstructured{}
+	err := u.UnmarshalJSON(rx.Raw)
+	if err != nil {
+		return nil, err
+	}
+	return &u, nil
+}
+
+func ensureCorrectGVK(obj runtime.Object, gvk schema.GroupVersionKind) error {
+	convertedGVK := obj.GetObjectKind().GroupVersionKind()
+	if convertedGVK != gvk {
+		return fmt.Errorf("invalid GVK returned by conversion webhook. expected=%s, actual=%s", gvk, convertedGVK)
+	}
+	return nil
+}
+
+func (c *webhookConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	if isEmptyUnstructuredObject(in) {
```

Comment:
when does this happen? Why should we handle that in a special way?

Reply:
actually now that we fixed the example webhook and also expect conversion webhooks to not fail on empty objects, this is less important. But it happens. This is where it initiated:

https://github.com/kubernetes/kubernetes/blob/dbb448bbdcb9e440eee57024ffa5f1698956a054/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L201

The config.Type comes from a `NewFunc` which creates an empty object and this line will pass that eventually to webhook converter to convert an empty object from one version to another.

I don't think we should call webhooks on an empty object as the behaviour should be like NopConverter. Generally conversion should not fail based on fields not exists. That is validation problem not conversion. Also we do not expect conversion webhook to do any defaulting too. So there is no point calling webhook on empty object.

Because we consistently create these empty objects in our machinery logic, it make sense to special case it here.

Reply:
It's some smoke test we added to the storage plumbing. @liggitt @deads2k do you remember why?

In theory an empty object in one version could map to a non-empty object in another version. So strictly speaking this special casing is wrong.

Reply:
'In theory an empty object in one version could map to a non-empty object in another version. So strictly speaking this special casing is wrong.'

how? do you have an example?